lib/haliax/src/haliax/partitioning.py

```diff
@@ -715,16 +716,127 @@ def round_axis_for_partitioning(axis: Axis, mapping: ResourceMapping | None = No
+    return _resolve_mesh()
 
-    warnings.warn(
-        "`_get_mesh` is deprecated; use `jax's get_abstract_mesh or get_concrete_mesh` instead",
-        DeprecationWarning,
-        stacklevel=2,
-    )
 
-    mesh = _resolve_mesh()
-    return mesh
+@typing.overload
+def shard_map(
+    f: Callable[Args, R],
+    *,
+    in_specs: Any = None,
+    out_specs: Any = None,
+    mesh: Mesh | None = None,
+    axis_mapping: ResourceMapping | None = None,
+    check_rep: bool = False,
+    **kwargs,
+) -> Callable[Args, R]: ...
+
+
+@typing.overload
+def shard_map(
+    *,
+    in_specs: Any = None,
+    out_specs: Any = None,
+    mesh: Mesh | None = None,
+    axis_mapping: ResourceMapping | None = None,
+    check_rep: bool = False,
+    **kwargs,
+) -> typing.Callable[[Callable[Args, R]], Callable[Args, R]]: ...
+
+
+def shard_map(
+    f: Optional[Callable] = None,
+    *,
+    in_specs: PyTree = None,
+    out_specs: PyTree = None,
+    mesh: Optional[Mesh] = None,
+    axis_mapping: Optional[ResourceMapping] = None,
+    check_rep: bool = False,
+    **shmap_kwargs: dict,
+) -> Callable:
+    """A NamedArray-friendly wrapper around :func:`jax.experimental.shard_map.shard_map`.
+
+    This function can be used either as ``haliax.shard_map(fn, ...)`` or as a
+    decorator::
+
+        @haliax.shard_map
+        def fn(x):
+            ...
+
+    Note that, unlike the JAX version, you don't need to provide in_specs or out_specs if all arguments and the return value
+    are ``NamedArray`` objects. The mesh can be inferred from the context mesh (which JAX could do, but doesn't).
+    Axis mapping can be inferred from the context axis mapping or provided explicitly.
+
+    Also note that this function will call ``f`` twice at trace time unless out_specs is provided: once to infer the output
+    pspec, and once to actually run the computation.
+
+    Args:
+        f: The function to apply with ``shard_map``.
+        in_specs: Optional PyTree prefix describing the input sharding. If None, the input
+            specifications are inferred from the input NamedArray arguments and the
+            provided `axis_mapping`.
+        out_specs: Like `in_specs` but for the output. If `None` the output
+            specifications are inferred by evaluating `f` on placeholder inputs
+            and using the returned axis names.
+        mesh: The mesh to run the computation on. Defaults to the current mesh
+            returned by [jax.sharding.get_abstract_mesh][].
+        axis_mapping: Optional mapping from logical axis names to mesh axis names
+            used when converting `Axis` objects to `PartitionSpec`.
+        check_rep: Passed through to `jax.shard_map`.
+        **shmap_kwargs: Additional arguments forwarded to `jax.shard_map`.
+
+    Returns:
+        A wrapped function that accepts and returns `NamedArray` objects
+        according to the provided specifications.
+    """
+
+    if f is None:
+        return lambda fn: shard_map(
```

Comment:
Just making sure I follow, this returns an generic shard_map wrapper for any function with the same type signature?

Reply:
yeah this is the "use it as a decorator or a higher order function" pattern